cdap-security/src/main/java/co/cask/cdap/security/securestore/FileSecureStoreProvider.java

```diff
@@ -0,0 +1,543 @@
+  private static final char[] SECURE_STORE_DEFAULT_PASSWORD = "cdapsecret".toCharArray();
+  /*
+   Java Keystore needs an algorithm name to store a key, it is not used for any checks, only stored,
+   since we are not handling the encryption we don't care about this.
+  */
+  private static final String ALGORITHM_PROXY = "none";
+
+  /*
+   A cache for metadata.
+   The assumption is that the number of entries won't be large enough to worry about memory.
+  */
+  private final Map<String, SecureStoreMetadata> cache = new HashMap<>();
+  private final char[] password;
+  private final Path path;
+  private final Lock readLock;
+  private final Lock writeLock;
+  private final CConfiguration cConf;
+
+  private KeyStore keyStore;
+  private boolean changed = false;
+
+  private FileSecureStoreProvider(CConfiguration cConf) {
+    this.cConf = cConf;
+    // Get the path to the keystore file
+    String pathString = cConf.get(Constants.Security.Store.SECURE_STORE_FILE_PATH);
+    if (pathString == null || pathString.isEmpty()) {
+      pathString = SECURE_STORE_DEFAULT_FILE_PATH;
+    }
+    Path dir = Paths.get(pathString);
+    path = dir.resolve(SECURE_STORE_DEFAULT_FILE_NAME);
+
+    // Get the keystore password
+    String passwordString = cConf.get(Constants.Security.Store.SECURE_STORE_FILE_PASSWORD);
+    if (passwordString == null || passwordString.isEmpty()) {
+      password = SECURE_STORE_DEFAULT_PASSWORD;
+    } else {
+      password = passwordString.toCharArray();
+    }
+
+    ReadWriteLock lock = new ReentrantReadWriteLock(true);
+    readLock = lock.readLock();
+    writeLock = lock.writeLock();
+  }
+
+  public static FileSecureStoreProvider getInstance(CConfiguration cConf) throws IOException {
+    FileSecureStoreProvider fileSecureStoreProvider = new FileSecureStoreProvider(cConf);
+    // Locate and load the keystore
+    fileSecureStoreProvider.locateKeystore();
+    return fileSecureStoreProvider;
+  }
+
+  /**
+   * Stores an element in the secure store. If the key already exists then delete it first.
+   * @param name
+   * @param data
+   * @param properties
+   * @throws IOException
+   */
+  void put(String name, byte[] data, Map<String, String> properties) throws IOException {
+    String metaKey = constructMetadataKey(name);
+    writeLock.lock();
+    try {
+      if (keyStore.containsAlias(name) || cache.containsKey(metaKey)) {
+        // Clear the existing key so that we can write the new one.
+        delete(name);
+      }
+      keyStore.setKeyEntry(name, new SecretKeySpec(data, ALGORITHM_PROXY),
+                           password, null);
+      SecureStoreMetadata meta = FileSecureStoreMetadata.of(name, properties);
+      /*
+        The data was written successfully to the key store. Now try to write the metddata.
+        If this fails then try to delete the data and throw an exception.
+       */
+      try {
+        keyStore.setKeyEntry(metaKey, new KeyMetadata(meta), password, null);
```

Comment:
Most of the time we would like to access them separately. Java keystore will need them to be serialized together to be stored. So, if we put them together we would need to get the data and deserialize both each time.